### PR TITLE
fix(harness): fetch full skill directory from URL base via ForgeClient

### LINKED ISSUE #2705: bug(harness): v0.22.0 URL-base skill resolution drops companion files, breaks review agent for all users
## What happens

Since the v0 tag was moved to v0.22.0 (2026-06-26 ~16:53 UTC), **all review agent runs fail** with `FAIL: output/agent-result.json not found` across every per-org `.fullsend` installation.

The root cause is PR #2690 (`feat(harness): resolve declarative resources from URL-referenced bases`). When skills are resolved from a URL-referenced base, only SKILL.md is fetched via `raw.githubusercontent.com`. Companion directories (`sub-agents/`, `meta-prompts/`, `scripts/`) are silently dropped.

The `pr-review` skill depends on `sub-agents/*.md` to dispatch parallel specialized reviewers. Without them, the agent receives a generic `"Run the agent task"` prompt, produces no output, and validation fails.

PR #2697 added a warning log for this case but did not prevent the breakage:
\`\`\`
! Base: skill "skills/pr-review" was fetched from a URL base with only SKILL.md;
  companion files (sub-agents, scripts, meta-prompts) may be missing
\`\`\`

## What should happen

Skills resolved from URL bases must include all companion files, or the harness should fail fast with a clear error when a multi-file skill is fetched with only SKILL.md. The v0 tag should not be moved to a release that has known-broken resource resolution for the default skill set.

## How to reproduce

1. Open any PR on a repo using a per-org `.fullsend` deployment with default `fullsend_ai_ref: v0`
2. Review agent triggers
3. Observe validation failure: `FAIL: output/agent-result.json not found`
4. Check logs for the `! Base: skill "skills/pr-review" was fetched from a URL base with only SKILL.md` warning on all 4 skills

## Context

- **Scope:** Only the review agent is broken. Code, fix, triage, and retro agents use skills that are SKILL.md-only (no companion files), so they are unaffected.
- **Blast radius:** All per-org `.fullsend` installations using `fullsend_ai_ref: v0` (the default) — fullsend-ai, konflux-ci, and every other enrolled org.
- **Timeline:**
  - v0.22.0 released and v0 tag moved: 2026-06-26 16:53 UTC
  - Last successful review run (fullsend-ai/.fullsend): 18:14 UTC
  - First failed review run: 19:46 UTC
  - All subsequent runs: failure (6+ consecutive on fullsend-ai/.fullsend alone)
- **Confirmed failing runs:**
  - fullsend-ai/.fullsend: 28261427923, 28261778104, 28262390410, 28262503589, 28263142411
  - konflux-ci/.fullsend: multiple runs queued/failing since 20:40 UTC
- **Immediate mitigation:** Roll back v0 tag to v0.21.0
- **Related PRs:**
  - #2690 — introduced URL-base resolution (merged 2026-06-26)
  - #2697 — added warning for SKILL.md-only fetch (merged 2026-06-26)
  - #2139 — earlier foundational work: model skills as directories (merged 2026-06-11)

### PR DESCRIPTION
**Pull request by waynesun09**
## Summary

- **Fixes #2705** — URL-base skill resolution in v0.22.0 drops companion files (sub-agents/, scripts/, meta-prompts/), breaking the review agent for all per-org installations
- Adds `ParseRawContentURL()` to parse `raw.githubusercontent.com` URLs into forge info (owner/repo/ref/path)
- Adds `ForgeClient` field to `ComposeOpts` so `fetchBaseSkill()` can use `ListDirectoryContents` + `GetFileContentAtRef` to fetch the full skill directory tree
- Requires ForgeClient (set GITHUB_TOKEN) or fails fast with a clear error directing users to run `fullsend lock` for offline pre-caching
- Adds empty-field validation to `ParseRawContentURL`
- Fixes ForgeClient injection in `lock.go` to match the `run.go` pattern
- Wires `ForgeClient` into `ComposeOpts` in both `run.go` and `lock.go` CLI callers

## Impact

Only the `pr-review` skill has companion files today (`meta-prompt.md`, `sub-agents/`). Other skills (`code-review`, `code-implementation`, `fix-review`, `issue-labels`, `retro-analysis`) are SKILL.md-only and are unaffected by this bug or this fix.

## Test plan

- [x] `ParseRawContentURL()` — 8 unit tests covering valid URLs, edge cases, wrong host, missing segments
- [x] `fetchBaseSkill()` with `FakeClient` — full-directory fetch, API error, missing SKILL.md, no ForgeClient error, parse error, audit error
- [x] Cache hit path — offline cache hit test preserved
- [x] CLI tests — lock and run tests updated to match new fail-fast behavior
- [x] `go build ./...` — compiles cleanly
- [x] `go test ./internal/forge/... ./internal/harness/... ./internal/cli/...` — all pass

### CODE REVIEW

File: internal/harness/compose.go [hunk 1]
```diff
@@ -807,6 +818,23 @@ func fetchBaseSkill(ctx context.Context, field, baseURLDir, skillPath string, al
...
+		if err == nil {
+			return dep, treePath, nil
+		}
+		// Log the failure and fall through to SKILL.md-only fetch so the
+		// agent can still run with degraded skill content.
+		if opts.AuditLogPath != "" {
+			_ = auditBaseFetch(opts, skillFileURL, "forge-client-error:"+err.Error(), allowedBy, false, time.Now().UTC(), "skill-dir-fetch-failed")
+		}
+	}
+
```
Comment (ggallen) [inline internal/harness/compose.go:823]: **Design: Remove the SKILL.md-only fallback entirely.**

This block silently swallows the forge API error and falls through to the old SKILL.md-only fetch. This is exactly the failure mode that caused the v0.22.0 outage — skills load but don't work because companions are missing, and the user only sees a warning buried in logs.

Instead:
- If `ForgeClient != nil` and the URL is forge-parseable: use the forge API. If it fails, **propagate the error**.
- If `ForgeClient == nil` and the URL is forge-parseable: error with "skill directories from forge URLs require a GitHub token".
- If the URL is not forge-parseable: error with "skill directories from non-forge URL bases are not supported".

One code path. No fallback. Skills always resolve as full directories or composition fails fast.

This also lets you delete `isSkillMDOnly()`, the conditional `cacheWarn` logic, and the entire SKILL.md-only fetch path below this block.

File: internal/cli/lock.go [hunk 2]
```diff
@@ -197,13 +197,22 @@ func lockOneAgent(ctx context.Context, agentName, absFullsendDir, forgeFlag stri
 	seen := make(map[string]bool)
 	linted := make(map[string]bool) // track reported lint diagnostics to avoid duplicates across forge variants
 
+	// Best-effort ForgeClient for base composition: enables full-directory
+	// skill fetches (sub-agents, meta-prompts). Nil is safe — fetchBaseSkill
+	// falls back to SKILL.md-only.
+	var composeForgeClient forge.Client
+	if token, tokenErr := resolveToken(); tokenErr == nil {
+		composeForgeClient = gh.New(token)
```
Comment (ggallen) [inline internal/cli/lock.go:204]: **Bug: Missing `rFlags.forgeClient` check.**

`run.go` checks `rFlags.forgeClient != nil` before falling back to `resolveToken()`, but this code skips that check. Lock tests cannot inject a mock forge client for compose-time skill resolution.

```suggestion
	var composeForgeClient forge.Client
	if rFlags.forgeClient != nil {
		composeForgeClient = rFlags.forgeClient
	} else if token, tokenErr := resolveToken(); tokenErr == nil {
		composeForgeClient = gh.New(token)
	}
```

File: internal/harness/compose.go [hunk 3]
```diff
@@ -807,6 +818,23 @@ func fetchBaseSkill(ctx context.Context, field, baseURLDir, skillPath string, al
 		return Dependency{}, "", fmt.Errorf("base %s: URL %s not in cache and offline mode is enabled (run 'fullsend lock' first)", field, skillFileURL)
 	}
 
+	// When ForgeClient is available, fetch the full skill directory via the
+	// GitHub Contents API. This retrieves companion files (sub-agents/,
+	// scripts/, meta-prompts/) that the SKILL.md-only path misses.
+	if opts.ForgeClient != nil {
+		dep, treePath, err := fetchBaseSkillDir(ctx, field, skillDirURL, skillFileURL, skillPath, allowedBy, opts)
+		if err == nil {
```
Comment (ggallen) [inline internal/harness/compose.go]: **Audit log corruption:** This writes `"forge-client-error:" + err.Error()` into the `SHA256` field of the audit entry. That field is semantically a hex-encoded hash — this puts arbitrary error text into it.

Moot if the fallback is removed (per the design comment). If kept, use a separate field or a sentinel value.

File: internal/forge/url.go [hunk 4]
```diff
@@ -92,6 +92,61 @@ func ParseForgeURL(rawURL string) (*ForgeURLInfo, error) {
...
+
+	owner := segments[0]
+	repo := segments[1]
+	ref := segments[2]
+
+	var repoPath string
+	if len(segments) > 3 {
+		repoPath = strings.Join(segments[3:], "/")
+	}
+
```
Comment (ggallen) [inline internal/forge/url.go:150]: **Missing validation:** `ParseForgeURL` validates that `owner` and `repo` are non-empty (lines 66-71). This function skips those checks. A URL like `https://raw.githubusercontent.com//repo/ref/path` would produce `ForgeURLInfo{Owner: ""}`, causing confusing downstream errors in `ListDirectoryContents`.

Also consider extracting shared helpers with `ParseForgeURL` (fragment stripping, scheme validation, segment splitting) — both functions duplicate the same structure.

File: internal/harness/compose.go [hunk 5]
```diff
@@ -1047,3 +1144,13 @@ func mergeForgeConfigInto(base, child *ForgeConfig) {
 		child.ValidationLoop = base.ValidationLoop
 	}
 }
+
+// isSkillMDOnly returns true if the directory contains only a SKILL.md file
+// (no companion files like sub-agents/, scripts/, meta-prompts/).
+func isSkillMDOnly(dir string) bool {
+	entries, err := os.ReadDir(dir)
+	if err != nil {
+		return true
```
Comment (ggallen) [inline internal/harness/compose.go]: **Wrong default on error:** If `os.ReadDir` fails (permissions, corruption), this returns `true` — suppressing the warning and hiding the problem. Moot if the fallback path is removed entirely.

**Review by ggallen — DISMISSED** [review]
The core fix is right — threading `ForgeClient` into `ComposeOpts` and using `ListDirectoryContents` + `GetFileContentAtRef` to fetch full skill directory trees. But the SKILL.md-only fallback needs to go. It preserves exactly the silent-degradation failure mode that caused the v0.22.0 outage. Skills should either fully resolve via the forge API or fail fast — one code path, no fallback.

Removing the fallback also lets you delete `isSkillMDOnly`, the conditional warning logic, and most of the old SKILL.md-only fetch path. See inline comments for specifics.

**Review by ralphbean — APPROVED** [review]
LGTM. The latest commit addresses all of the prior review feedback — fallback removed, allowlist check added, lock.go wiring fixed, validation added, dead code cleaned up.
